src/librbd/ImageCtx.h

```diff
@@ -68,6 +68,7 @@ namespace librbd {
                                         // a format librados can understand
     std::map<librados::snap_t, SnapInfo> snap_info;
     std::map<std::string, librados::snap_t> snap_ids;
+    std::map<cls::rbd::SnapshotNamespace, librados::snap_t> snap_namespace_ids;
```

Comment:
I tried merging snap_ids and snap_namespace_ids when I just started working on this and it required so much refactoring of the existing code that I decided to do it in a separate PR if at all. Here the issues I've encountered. Probably, you'll be able to repel my fears. If I use pair <SnapshotNamespace, String> as the key for each entry then I can't search by name only? It means that everywhere where we want to determine the id of a snapshot by its name we need to specify its namespace as well. It means changing even command line tools in order to allow the user to specify what namespace he wants to look in. Which seems to me like a too big change.

If we use reversed pair <string, SnapshotNamespace> then we can look for snapshots using only their names, but then we can't look for snapshots using only their Namespace. But this is why we want this index in the first place. Taking all of this into account I decided to just have a separate index for namespaces. What is your thought on this?